library/core/src/str/validations.rs

```diff
@@ -112,172 +112,308 @@ where
     Some(ch)
 }
 
-const NONASCII_MASK: usize = usize::repeat_u8(0x80);
+// The transition table of shift-based DFA for UTF-8 validation.
+// Ref: <https://gist.github.com/pervognsen/218ea17743e1442e59bb60d29b1aa725>
+//
+// In short, we encode DFA transitions in an array `TRANS_TABLE` such that:
+// ```
+// TRANS_TABLE[next_byte] =
+//     (target_state1 * BITS_PER_STATE) << (source_state1 * BITS_PER_STATE) |
+//     (target_state2 * BITS_PER_STATE) << (source_state2 * BITS_PER_STATE) |
+//     ...
+// ```
+// Thanks to pre-multiplication, we can execute the DFA with one statement per byte:
+// ```
+// let state = initial_state * BITS_PER_STATE;
+// for byte in .. {
+//     state = TRANS_TABLE[byte] >> (state & ((1 << BITS_PER_STATE) - 1));
+// }
+// ```
+// By choosing `BITS_PER_STATE = 6` and `state: u64`, we can replace the masking by `wrapping_shr`.
+// ```
+// // shrx state, qword ptr [table_addr + 8 * byte], state   # On x86-64-v3
+// state = TRANS_TABLE[byte].wrapping_shr(state);
+// ```
+//
+// On platform without 64-bit shift, especially i686, we split the `u64` next-state into
+// `[u32; 2]`, and each `u32` stores 5 * BITS_PER_STATE = 30 bits. In this way, state transition
+// can be done in only 32-bit shifts and a conditional move, which is several times faster
+// (in latency) than ordinary 64-bit shift (SHRD).
+//
+// The DFA is directly derived from UTF-8 syntax from the RFC3629:
+// <https://datatracker.ietf.org/doc/html/rfc3629#section-4>.
+// We assign S0 as ERROR and S1 as ACCEPT. DFA starts at S1.
+// Syntax are annotated with DFA states in angle bracket as following:
+//
+// UTF8-char   = <S1> (UTF8-1 / UTF8-2 / UTF8-3 / UTF8-4)
+// UTF8-1      = <S1> %x00-7F
+// UTF8-2      = <S1> %xC2-DF                <S2> UTF8-tail
+// UTF8-3      = <S1> %xE0                   <S3> %xA0-BF <S2> UTF8-tail /
+//               <S1> (%xE1-EC / %xEE-EF)    <S4> 2( UTF8-tail ) /
+//               <S1> %xED                   <S5> %x80-9F <S2> UTF8-tail
+// UTF8-4      = <S1> %xF0    <S6> %x90-BF   <S4> 2( UTF8-tail ) /
+//               <S1> %xF4    <S7> %x80-8F   <S4> 2( UTF8-tail ) /
+//               <S1> %xF1-F3 <S8> UTF8-tail <S4> 2( UTF8-tail )
+//
+// UTF8-tail   = %x80-BF   # Inlined into above usages.
+const BITS_PER_STATE: u32 = 6;
+const STATE_MASK: u32 = (1 << BITS_PER_STATE) - 1;
+const STATE_CNT: usize = 9;
+#[allow(clippy::all)]
+const ST_ERROR: u32 = 0 * BITS_PER_STATE as u32;
+#[allow(clippy::all)]
+const ST_ACCEPT: u32 = 1 * BITS_PER_STATE as u32;
 
-/// Returns `true` if any byte in the word `x` is nonascii (>= 128).
+/// Platforms that does not have efficient 64-bit shift and should use 32-bit shift fallback.
+const USE_SHIFT32: bool = cfg!(all(
+    any(target_pointer_width = "16", target_pointer_width = "32"),
+    // WASM32 supports 64-bit shift.
+    not(target_arch = "wasm32"),
```

Comment:
You maybe want to include x86_64 and aarch64 here, which have targets with 32-bit pointers.

Reply:
> You maybe want to include x86_64 and aarch64 here, which have targets with 32-bit pointers.

How can you target x86_64 and aarch64 but forcing 32-bit pointers? I don't know it's possible.

Reply:
By using an ABI with 32 bit pointers, such as [x32](https://en.wikipedia.org/wiki/X32_ABI). I don't think any of those are popular today (certainly not tier 1-2 targets in rustc) but they do exist.